src/CCTPBase.sol

```diff
@@ -0,0 +1,169 @@
+
+pragma solidity ^0.8.13;
+
+import "./interfaces/IWormholeReceiver.sol";
+import "./interfaces/IWormholeRelayer.sol";
+import "./interfaces/ITokenBridge.sol";
+import {IERC20} from "./interfaces/IERC20.sol";
+import "./interfaces/CCTPInterfaces/ITokenMessenger.sol";
+import "./interfaces/CCTPInterfaces/IMessageTransmitter.sol";
+
+import "./Utils.sol";
+import "./TokenBase.sol";
+
+library CCTPMessageLib {
+    uint8 constant CCTP_KEY_TYPE = 2;
+
+    // encoded using abi.encodePacked(domain, nonce)
+    struct CCTPKey {
+        uint32 domain;
+        uint64 nonce;
+    }
+
+    // encoded using abi.encode(message, signature)
+    struct CCTPMessage {
+        bytes message;
+        bytes signature;
+    }
+}
+
+abstract contract CCTPBase is TokenBase {
+    ITokenMessenger immutable circleTokenMessenger;
+    IMessageTransmitter immutable circleMessageTransmitter;
+    address immutable USDC;
+
+    constructor(
+        address _wormholeRelayer,
+        address _tokenBridge,
+        address _wormhole,
+        address _circleMessageTransmitter,
+        address _circleTokenMessenger,
+        address _USDC
+    ) TokenBase(_wormholeRelayer, _tokenBridge, _wormhole) {
+        circleTokenMessenger = ITokenMessenger(_circleTokenMessenger);
+        circleMessageTransmitter = IMessageTransmitter(_circleMessageTransmitter);
+        USDC = _USDC;
+    }
+
+    function getCCTPDomain(uint16 chain) internal pure returns (uint32) {
+        if (chain == 2) {
+            return 0;
+        } else if (chain == 6) {
+            return 1;
+        } else if (chain == 23) {
+            return 3;
+        } else if (chain == 24) {
+            return 2;
+        } else {
+            revert("Wrong CCTP Domain");
+        }
+    }
+
+    function redeemUSDC(bytes memory cctpMessage) internal returns (uint256 amount) {
+        (bytes memory message, bytes memory signature) = abi.decode(cctpMessage, (bytes, bytes));
+        uint256 beforeBalance = IERC20(USDC).balanceOf(address(this));
+        circleMessageTransmitter.receiveMessage(message, signature);
+        return IERC20(USDC).balanceOf(address(this)) - beforeBalance;
+    }
+}
+
+abstract contract CCTPSender is CCTPBase {
+    uint8 internal constant CONSISTENCY_LEVEL_FINALIZED = 15;
```

Comment:
https://book.wormhole.com/wormhole/3_coreLayerContracts.html?highlight=consiste#consistency-levels 

I think it could be anything other than 200/201 except for BSC where it actually needs to be >= 15? not certain, but I believe this is why 15 is used